src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.ClientCertProvider.cs

```diff
@@ -0,0 +1,255 @@
+                    GCHandle handle = GCHandle.FromIntPtr(certInfoPtr);
+                    certificateInfo = (CurlCbCertificateInfo)handle.Target;
+                }
+                catch (InvalidCastException)
+                {
+                    Debug.Fail("CurlCbCertificateInfo wasn't the GCHandle's Target");
+                    return CURLcode.CURLE_ABORTED_BY_CALLBACK;
+                }
+                catch (InvalidOperationException)
+                {
+                    Debug.Fail("Invalid GCHandle in CurlSslCallback");
+                    return CURLcode.CURLE_ABORTED_BY_CALLBACK;
+                }
+                Debug.Assert(certificateInfo != null, "Expected non-null CurbCbCertificateInfo in GCHandle");
+
+                int retCode = Interop.libssl.SSL_CTX_use_certificate(sslContext, certificateInfo._CertificateHandle);
+                if (retCode != 1)
+                {
+                    VerboseTrace("cannot use certificate ");
+                    return CURLcode.CURLE_ABORTED_BY_CALLBACK;
+                }
+
+                retCode = Interop.libssl.SSL_CTX_use_PrivateKey(sslContext, certificateInfo._PrivateKeyHandle);
+                if (retCode != 1)
+                {
+                    VerboseTrace("Failed to set private key");
+                    return CURLcode.CURLE_ABORTED_BY_CALLBACK;
+                }
+
+                retCode = Interop.libssl.SSL_CTX_check_private_key(sslContext);
+                if (retCode != 1)
+                {
+                    VerboseTrace("SET Cert failed");
+                    return CURLcode.CURLE_ABORTED_BY_CALLBACK;
+                }
+                VerboseTrace("check return code " + retCode);
+                VerboseTrace("successfully set client certificate");
+                return CURLcode.CURLE_OK;
+            }
+
+            private static string GeneratePassword()
+            {
+                byte[] bytes = new byte[SecurePasswordByteLength];
+                var prng = new Random();
+                prng.NextBytes(bytes);
+                return Convert.ToBase64String(bytes);
+            }
+
+            private static string GetCommonName(X509Certificate2 cert)
+            {
+                string subjectName = cert.Subject;
+                if (string.IsNullOrEmpty(subjectName))
+                {
+                    return string.Empty;
+                }
+
+                int begin = 0;
+                bool found = false;
+                for (int i=0; i< subjectName.Length - 3; i++)
+                {
+                    if (subjectName[i] == 'C' && subjectName[i+1] == 'N' && subjectName[i+2] == '=')
+                    {
+                        begin = i+3;
+                        found = true;
+                        break;
+                    }
+                }
+
+                if(!found)
+                {
+                    // Treat the whole name as Common Name
+                    return subjectName;
```

Comment:
No this is not defined anywhere. ```CN=``` prefix seems to be an openssl thing. I am not sure. For darwinssl and nss, the search in the respective stores failed when I do not try to strip the prefix. 

But there are other certs (admittedly those which I created using ```openssl```) where there were no subject-names. A little like below:

```OU=Blah Org, O=Contosso Pvt Ltd, L=Hyderabad, ST=AP, C=IN``` 

And for such certs it is going to fail no matter what I do. So I figured we could as well use the whole thing and leave it to the ssl backend to make use of it the way it sees fit.